File: C#/1_Creational Patterns/3_Abstract Factory/VehicleManufacturingSystem.cs
// Abstract Product
public interface IVehicle {
    void Assemble();
}

// Concrete Products
public class ElectricCar : IVehicle {
    public void Assemble() {
        Console.WriteLine("Assembling electric car...");
    }
}

public class GasolineCar : IVehicle {
    public void Assemble() {
        Console.WriteLine("Assembling gasoline car...");
    }
}

public class ElectricMotorcycle : IVehicle {
    public void Assemble() {
        Console.WriteLine("Assembling electric motorcycle...");
    }
}

public class GasolineMotorcycle : IVehicle {
    public void Assemble() {
        Console.WriteLine("Assembling gasoline motorcycle...");
    }
}

// Abstract Factory
public interface IVehicleFactory {
    IVehicle CreateVehicle();
}

// Concrete Factories
public class ElectricCarFactory : IVehicleFactory {
    public IVehicle CreateVehicle() {
        return new ElectricCar();
    }
}

public class GasolineCarFactory : IVehicleFactory {
    public IVehicle CreateVehicle() {
        return new GasolineCar();
    }
}

public class ElectricMotorcycleFactory : IVehicleFactory {
    public IVehicle CreateVehicle() {
        return new ElectricMotorcycle();
    }
}

public class GasolineMotorcycleFactory : IVehicleFactory {
    public IVehicle CreateVehicle() {
        return new GasolineMotorcycle();
    }
}

// Client
public class VehicleManufacturingSystem {
    static void Main(string[] args) {
        IVehicleFactory electricCarFactory = new ElectricCarFactory();
        IVehicleFactory gasolineCarFactory = new GasolineCarFactory();
        IVehicleFactory electricMotorcycleFactory = new ElectricMotorcycleFactory();
        IVehicleFactory gasolineMotorcycleFactory = new GasolineMotorcycleFactory();

        IVehicle electricCar = electricCarFactory.CreateVehicle();
        IVehicle gasolineCar = gasolineCarFactory.CreateVehicle();
        IVehicle electricMotorcycle = electricMotorcycleFactory.CreateVehicle();
        IVehicle gasolineMotorcycle = gasolineMotorcycleFactory.CreateVehicle();

        electricCar.Assemble();
        gasolineCar.Assemble();
        electricMotorcycle.Assemble();
        gasolineMotorcycle.Assemble();
    }
}

File: Python/1_Creational Patterns/3_Abstract Factory/VehicleManufacturingSystem.py
from abc import ABC, abstractmethod

# Abstract Product
class Vehicle(ABC):
    @abstractmethod
    def assemble(self):
        pass

# Concrete Products
class ElectricCar(Vehicle):
    def assemble(self):
        print("Assembling electric car...")

class GasolineCar(Vehicle):
    def assemble(self):
        print("Assembling gasoline car...")

class ElectricMotorcycle(Vehicle):
    def assemble(self):
        print("Assembling electric motorcycle...")

class GasolineMotorcycle(Vehicle):
    def assemble(self):
        print("Assembling gasoline motorcycle...")

# Abstract Factory
class VehicleFactory(ABC):
    @abstractmethod
    def create_vehicle(self):
        pass

# Concrete Factories
class ElectricCarFactory(VehicleFactory):
    def create_vehicle(self):
        return ElectricCar()

class GasolineCarFactory(VehicleFactory):
    def create_vehicle(self):
        return GasolineCar()

class ElectricMotorcycleFactory(VehicleFactory):
    def create_vehicle(self):
        return ElectricMotorcycle()

class GasolineMotorcycleFactory(VehicleFactory):
    def create_vehicle(self):
        return GasolineMotorcycle()

# Client
def main():
    electric_car_factory = ElectricCarFactory()
    gasoline_car_factory = GasolineCarFactory()
    electric_motorcycle_factory = ElectricMotorcycleFactory()
    gasoline_motorcycle_factory = GasolineMotorcycleFactory()

    electric_car = electric_car_factory.create_vehicle()
    gasoline_car = gasoline_car_factory.create_vehicle()
    electric_motorcycle = electric_motorcycle_factory.create_vehicle()
    gasoline_motorcycle = gasoline_motorcycle_factory.create_vehicle()

    electric_car.assemble()
    gasoline_car.assemble()
    electric_motorcycle.assemble()
    gasoline_motorcycle.assemble()

if __name__ == "__main__":
    main()
